src/Microsoft.OpenApi/Reader/OpenApiModelFactory.cs

```diff
@@ -193,146 +194,172 @@ public static T Parse<T>(string input,
-        /// <param name="diagnostic">Returns diagnostic object containing errors detected during parsing.</param>
-        /// <param name="settings">The OpenApiReader settings.</param>
-        /// <returns>Instance of newly created IOpenApiElement.</returns>
-        /// <returns>The OpenAPI element.</returns>
-        public static T Load<T>(string url, OpenApiSpecVersion version, out OpenApiDiagnostic diagnostic, OpenApiReaderSettings settings = null) where T : IOpenApiElement
+        private static async Task<ReadResult> InternalLoadAsync(Stream input, string format, OpenApiReaderSettings settings, CancellationToken cancellationToken = default)
         {
-            var format = GetFormat(url);
-            settings ??= new OpenApiReaderSettings();
+            var reader = OpenApiReaderRegistry.GetReader(format);
+            var readResult = await reader.ReadAsync(input, settings, cancellationToken);
 
-#pragma warning disable VSTHRD002 // Avoid problematic synchronous waits
-            var stream = GetStreamAsync(url).GetAwaiter().GetResult();
-#pragma warning restore VSTHRD002 // Avoid problematic synchronous waits
+            if (settings.LoadExternalRefs)
+            {
+                var diagnosticExternalRefs = await LoadExternalRefsAsync(readResult.OpenApiDocument, cancellationToken, settings, format);
+                // Merge diagnostics of external reference
+                if (diagnosticExternalRefs != null)
+                {
+                    readResult.OpenApiDiagnostic.Errors.AddRange(diagnosticExternalRefs.Errors);
+                    readResult.OpenApiDiagnostic.Warnings.AddRange(diagnosticExternalRefs.Warnings);
+                }
+            }
 
-            return Load<T>(stream, version, format, out diagnostic, settings);
+            return readResult;
         }
 
-        /// <summary>
-        /// Reads the stream input and parses the fragment of an OpenAPI description into an Open API Element.
-        /// </summary>
-        /// <typeparam name="T"></typeparam>
-        /// <param name="input">Stream containing OpenAPI description to parse.</param>
-        /// <param name="version">Version of the OpenAPI specification that the fragment conforms to.</param>
-        /// <param name="format"></param>
-        /// <param name="diagnostic">Returns diagnostic object containing errors detected during parsing.</param>
-        /// <param name="settings">The OpenApiReader settings.</param>
-        /// <returns>Instance of newly created IOpenApiElement.</returns>
-        /// <returns>The OpenAPI element.</returns>
-        public static T Load<T>(Stream input, OpenApiSpecVersion version, string format, out OpenApiDiagnostic diagnostic, OpenApiReaderSettings settings = null) where T : IOpenApiElement
+        private static async Task<OpenApiDiagnostic> LoadExternalRefsAsync(OpenApiDocument document, CancellationToken cancellationToken, OpenApiReaderSettings settings, string format = null)
```

Comment:
nit: move cancellation token parameter to the last, make it optional to follow conventions